core/src/main/java/org/testcontainers/containers/GenericContainer.java

```diff
@@ -577,21 +538,41 @@ public Integer getMappedPort(final int originalPort) {
         }
     }
 
+    /**
+     * <b>Resolve</b> Docker image and set it.
+     *
+     * @param dockerImageName image name
+     */
+    @SneakyThrows
     public void setDockerImageName(@NonNull String dockerImageName) {
+        String[] parts = dockerImageName.split(":");
+        boolean isTagNamePresent = parts.length == 2;
+        Preconditions.checkArgument(isTagNamePresent, "No image tag was specified in docker image name (" + dockerImageName + "). Please provide a tag; this may be 'latest' or a specific version");
 
-        this.dockerImageName = dockerImageName;
+        RemoteDockerImage remoteDockerImage = new RemoteDockerImage(parts[0], parts[1]);
 
-        boolean isTagNamePresent = dockerImageName.split(":").length == 2;
-        Preconditions.checkArgument(isTagNamePresent, "No image tag was specified in docker image name (" + dockerImageName + "). Please provide a tag; this may be 'latest' or a specific version");
+        try {
+            // Mimic old behavior where we resolve image once it's set
+            remoteDockerImage.get();
+        } catch (ExecutionException e) {
+            throw e.getCause();
+        }
 
-        Profiler profiler = new Profiler("Rule creation - prefetch image");
-        profiler.setLogger(logger());
+        this.image = remoteDockerImage;
+    }
+
+    /**
+     * Get image name.
+     *
+     * @return image name
+     */
+    @NonNull
+    @SneakyThrows
```

Comment:
@SneakyThrows from Lombok here to make it behave as previously - if exception was thrown during pull, it will be thrown here as well.